### PR TITLE
[#8505] Run PyPy with Travis-CI

### PR DESCRIPTION
**Pull request by adiroiban**
See https://twistedmatrix.com/trac/ticket/8505

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -25,7 +38,23 @@ env:
...
+          pushd "$PYENV_ROOT" && git pull && popd
+        else
+          rm -rf "$PYENV_ROOT" && git clone --depth 1 https://github.com/yyuu/pyenv.git "$PYENV_ROOT"
+        fi
+        export PYPY_VERSION="5.3"
+        "$PYENV_ROOT/bin/pyenv" install --skip-existing "pypy-$PYPY_VERSION"
+        virtualenv --python="$PYENV_ROOT/versions/pypy-$PYPY_VERSION/bin/python" "$HOME/virtualenvs/pypy-$PYPY_VERSION"
+        source "$HOME/virtualenvs/pypy-$PYPY_VERSION/bin/activate"
+      fi
   - pip install tox tox-travis
```
Comment (ryanhiebert) [inline .travis.yml]: You don't technically need to install `tox` explicitly, since `tox-travis` depends on `tox`... :-)

File: .travis.yml [hunk 2]
```diff
@@ -15,17 +26,27 @@ cache:
...
+        export PYENV_ROOT="$HOME/.pyenv"
+        if [ -f "$PYENV_ROOT/bin/pyenv" ]; then
+          pushd "$PYENV_ROOT" && git pull && popd
+        else
+          rm -rf "$PYENV_ROOT" && git clone --depth 1 https://github.com/yyuu/pyenv.git "$PYENV_ROOT"
+        fi
+        export PYPY_VERSION="5.3"
+        "$PYENV_ROOT/bin/pyenv" install --skip-existing "pypy-$PYPY_VERSION"
+        virtualenv --python="$PYENV_ROOT/versions/pypy-$PYPY_VERSION/bin/python" "$HOME/virtualenvs/pypy-$PYPY_VERSION"
+        source "$HOME/virtualenvs/pypy-$PYPY_VERSION/bin/activate"
```
Comment (markrwilliams) [inline .travis.yml]: @adiroiban while i investigate the coverage-related `RuntimeError`, you can use a crummy hack to work around it: increase the recursion limit.  10,000 seems to work on my laptop.
here's one way to do that:

```
echo "import sys; sys.setrecursionlimit(10000)" > $HOME/virtualenvs/pypy-$PYPY_VERSION/
/lib-python/2.7/sitecustomize.py
```

I hope I can figure out a more satisfactory solution!

Comment (adiroiban) [inline .travis.yml]: Thanks. I will give it a try... but first I will use pypy 5.1 as coverage is not documented to work with pypy 5.3 http://coverage.readthedocs.io/en/latest/

again... I have never used pypy so I have no idea about all these crazy version numbers in pypy and pypy3 :) ... and in pyenv I see 5.1 and 5.1.1  :)

**Comment by markrwilliams** [comment]
@adiroiban it looks like travis did not to run on the last two commits -- was that intentional??

**Comment by adiroiban** [comment]
> @adiroiban it looks like travis did not to run on the last two commits -- was that intentional??

Nope. Something is broken with Travis-CI as it was ignoring my push commits.

**Comment by markrwilliams** [comment]
@adiroiban fair enough.  i'm happy to help if i can!

**Comment by adiroiban** [comment]
I have other PR and travis-ci is not triggered ... so I need to investigate ...I have sent an email to Travis-CI support team

anyway we need to fix the coverage issue and this will also need to be deployed on our buildbot env

File: .travis.yml [hunk 3]
```diff
@@ -15,17 +26,29 @@ cache:
...
+        echo "import sys; sys.setrecursionlimit(10000)" > $HOME/virtualenvs/pypy-$PYPY_VERSION/
+/lib-python/2.7/sitecustomize.py
```
Comment (BanzaiMan) [inline .travis.yml]: What is this line supposed to do? This line introduces YAML syntax error, so Travis CI build will not run.

**Comment by markrwilliams** [comment]
@adiroiban so it didn't _fix_ the issue but it did seem to let the code advance a little farther.  if you're willing to give it another shot, you can bump the recursion limit to 10000.  otherwise i'll report back after i've gotten a chance to run coverage & trial under a debug build of pypy.

**Comment by adiroiban** [comment]
> if you're willing to give it another shot, you can bump the recursion limit to 10000. otherwise i'll report back after i've gotten a chance to run coverage & trial under a debug build of pypy.

The limit is already at 10.000 ... as I have just copy pasted from your comment :)

Maybe is worth getting in touch with coverage people to see if something special is needed for PyPy

**Comment by markrwilliams** [comment]
@adiroiban sorry, i meant 100,000 - github ate my comment and i hastily retyped it.

i can reach out to the coverage people, but i'm pretty sure this is a pypy specific issue :( i'll update this PR when i know more.

**Comment by rodrigc** [comment]
@markrwilliams: take a look at https://travis-ci.org/twisted/twisted/jobs/142311357

**Comment by markrwilliams** [comment]
@rodrigc Sorry, I should have mentioned that I've filed an issue with PyPy about this: https://bitbucket.org/pypy/pypy/issues/2335/maximum-recursion-depth-exceeded-with
I'll update this PR when a solution is found.

**Comment by rodrigc** [comment]
@markrwilliams How many Pypy patches for Twisted do you have?
Can you extract a subset of those changes which do **not** require `if PYPY`, and
can run completely on Python 2.7, and pass the coverage tests on Python 2.7?
If we can get a subset of those patches committed into Twisted, that will allow some forward progress.

Getting blocked because the coverage tool itself doesn't work under Pypy is :(

**Comment by adiroiban** [comment]
The codecov patch protection was lifted see https://github.com/twisted-infra/braid/issues/213

so you can no merge pypy patches even if their coverage is not reported

**Comment by markrwilliams** [comment]
Armin Rigo [just pushed a fix](https://bitbucket.org/pypy/pypy/issues/2335/maximum-recursion-depth-exceeded-with#comment-29086607) for the coverage issue.  Looks like we can revist it this after the next PyPy release!

**Comment by adiroiban** [comment]
I have pushed a version with 100.000 recursion limit... but not luck.... so I am reverting that change.

Rather than having this code unmerged, I have pushed it for review as at least it can run pypy on travis ci... once a new pypy release is ready, in theory we should only update the version in the travis install.sh

When is the new pypy version planned to be released?

I had a quick look at the pypy website, but the release notes are done without any date ... very strange.

Similar with the bad habit that we have in Twisted, when a ticket is closed as fixed there is no note about when it will be released ... or in which version it was released

**Comment by glyph** [comment]
@adiroiban - can you update your branch please?  hopefully the push of the merge commit will also fix whatever broke the documentation buildbot…

**Comment by adiroiban** [comment]
@glyph updated. thanks!

**Comment by glyph** [comment]
Updating again in the hopes that the CI gods will be with us this time…

**Comment by glyph** [comment]
Ugh. They were, but now conflicts need to be resolved.

**Comment by glyph** [comment]
@adiroiban - the build is failing because the topfile is in the wrong place. Move it into `src/` and I think we can merge this…

File: .travis/install.sh [hunk 4]
```diff
@@ -1,20 +1,54 @@
...
+#
+# Activate the virtualenv if required.
+#
+if [ -f ~/.venv/bin/activate ]; then
+    # Initialize the virtualenv created at install time.
     source ~/.venv/bin/activate
 fi
 
 # Temporary workaround for https://github.com/pypa/setuptools/issues/776;
-# install (and thereby cache a built wheel of) cryptography.  (NB: We're
```
Comment (glyph) [inline .travis/install.sh]: Why delete this comment?  Seems irrelevant to the change in question.

File: .travis/run.sh [hunk 5]
```diff
@@ -2,8 +2,10 @@
 set -e
 set -x
 
-if [[ "$(uname -s)" == "Darwin" ]]; then
-    # Initialize the virtualenv created at install time.
```
Comment (glyph) [inline .travis/run.sh:6]: This change also seems spurious…

File: .travis.yml [hunk 6]
```diff
@@ -19,7 +19,22 @@ branches:
...
+    - PYPY_VERSION=pypy-5.4.1
+    # Some python versions (like PYPY) are installed using pyenv and we cache
+    # the compiled version.
+    - PYENV_ROOT="$HOME/.pyenv"
 
+addons:
+  apt:
+    packages:
+    - libssl-dev
+    - libssl1.0.0
```
Comment (glyph) [inline .travis.yml]: Obviously this is fine because the build works, but… why don't we need `libffi-dev`?

File: .travis.yml [hunk 7]
```diff
@@ -19,7 +19,22 @@ branches:
...
+
+cache:
+  directories:
+    - $HOME/.cache/pip
+    - $HOME/.pyenv
```
Comment (glyph) [inline .travis.yml]: Just curious - can this not be expressed as just `$PYENV_ROOT`?

File: .travis.yml (hunk 6 above)
Comment (adiroiban) [reply .travis.yml]: I don't know... but this is what we have in trunk.

File: .travis.yml (hunk 7 above)
Comment (adiroiban) [reply .travis.yml]: it can. done.

File: .travis/install.sh (hunk 4 above)
Comment (adiroiban) [reply .travis/install.sh]: I have removed it as for pypy this is no longer valid.

In travis we have pypy 2.5 and in venv we have pypy 5.4

File: .travis/run.sh (hunk 5 above)
Comment (adiroiban) [reply .travis/run.sh:6]: For pypy we also need to source the venv...

so instead of adding a condition for the OS / Travis ENV, I went for sourcing anything which was created at the install time.

In this way, if later we add a PYPY3 for example in install.sh we don't need to add a new source condition here in run.sh

**Comment by adiroiban** [comment]
Thanks @glyph for the review. I have moved to src and also updated the travis cache functionality 

It would be nice to have a command in pip which can remove packages which are not used for a long time.

for pyenv, I went for this simple pyenv uninstall command, but we might need a .travis/pyenv.cleanup.sh script which will remove any version, other than PYPY_VERSION ... and in the future we might want to keep other versions if pyenv

**Comment by adiroiban** [comment]
we are caching on travis the full ~/.cache/pip , but maybe we should only cache ~/.cache/pip/wheels

**Comment by njsmith** [comment]
Instead of doing all this work with pyenv and caching, wouldn't it be easier to install squeaky's portable pypy from scratch on each build? (The travis cache has to be downloaded to the build worker every time anyway, so it's only really useful for artifacts that you're building on travis.)

**Comment by glyph** [comment]
@njsmith - An alternate PR would be appreciated, but given that "all this work" is already done, is there much benefit to getting rid of it?

**Comment by njsmith** [comment]
I don't know. I just happened to glance at this out of curiosity and on a
quick skim it looked like there was some struggle with getting the cache
state management to work correctly -- the advantage of my suggestion is
that it'd be stateless.

**Comment by markrwilliams** [comment]
@adiroiban I agree with @njsmith and would prefer to use [squeaky's portable PyPy](https://github.com/squeaky-pl/portable-pypy).  I can take up this PR if that approach seems OK to you.

**Comment by adiroiban** [comment]
@markrwilliams go for it. Thanks!

I have never used pypy so my work from here was just trying to pay it forward for all the good things I use in Twisted basically great python 2.7 as I don't care for the rest :)

I think that I have just checked what Travis-CI was doing and I went on those lines.

**Comment by ulidtko** [comment]
*/me zaps a blessed wand of dead turning on the "Twisted on PyPy" thread. The PR rises from the dead.*

**Comment by glyph** [comment]
@ulidtko You have successfully nerd-sniped me.

**Comment by Julian** [comment]
What's wrong with the travis-provided PyPy that means pyenv needs to be involved?

**Comment by glyph** [comment]
Historically it has been unspeakably ancient, and only the most recent pypys are really worth testing. If travis is on the latest we could revert that part.

**Comment by glyph** [comment]
(I’d rather just land this as-is and do that as a follow-up though)

**Comment by Julian** [comment]
Got it -- yup sounds good (me personally I've [not had any issues recently
with it being terribly out of date](
https://travis-ci.org/Julian/jsonschema/jobs/569829619#L273), but yeah
think doing it as a follow up is fine)

File: src/twisted/python/compat.py [hunk 8]
```diff
@@ -852,6 +852,37 @@ def _get_async_param(isAsync=None, **kwargs):
...
+        passproto = [proto] * (proto is not None)
+        flags = fcntl(fd, F_GETFL)
+        try:
+            return realFromFD(fd, family, type, *passproto)
+        finally:
+            fcntl(fd, F_SETFL, flags)
+    realFromFD = socket.fromfd
+    if realFromFD.__name__ == fromFDWithoutModifyingFlags.__name__:
+        return
+    socket.fromfd = fromFDWithoutModifyingFlags
```
Comment (Julian) [inline src/twisted/python/compat.py:878]: So... personally I have experienced cases where twisted has monkey patched globally a thing and that monkey patch *introduced* bugs in non-twisted code...

I realize how tempting this is over instead just changing code to use a level of indirection, since I presume fromfd is reasonably widely used but... severe :(

Tell me to go away though I guess, and that this is the best way.

**Review by Julian — APPROVED** [review]
I really dislike the monkey patching, but other than that, lgtm.

File: src/twisted/python/compat.py (hunk 8 above)
Comment (glyph) [reply src/twisted/python/compat.py:878]: This is the best way, at least in this specific case.

- It's only monkeypatched on pypy3
- The behavior on pypy3 is buggy as heck and this will fix it for other programs as well.
- The monkeypatch is harmless (even if the monkeypatch were to get invoked on a different cpython, it would at worst be slightly inefficient unless someone is doing something *truly* batty with threads).
- I expect this to be relatively short-lived; it links to the bug, the bug is super straightforward and has a reproducer that's only a half a dozen lines.  The other bug I filed today, Armin has already fixed.
